### DISCUSSION TITLE
How to stub or spy on a an inner function

### DISCUSSION THREAD

**Discussion by Gmiliuta** [cypress-io/cypress#25897] [Questions and Help]
Hello everyone, 

We are using cypress for unit testing and we have a case where we need to stub/ spy an inner function as in example below.

```
// storage.js module

export function addDetails({ name, lastname }) {
      const allDetails =  JSON.parse(localStorage.getItem("personDetails"))
      allDetails.push(`${name} {lastname}`)
      
      localStorage.setItem("personDetails",  JSON.stringify(fullname))
      
      updateDetailsCount(allDetails)
}

export function updateDetailsCount(details) {
      // doing something with details
}

```

```
// stroage.cy.js unit test

import * as storageModule from './storage.js'

describe('addDetails', () => {
 
    it('should', () => {
        const name = 'tom';
        const lastname = 'jones'

        const updateDetailsCountSpy = cy.spy(storageModule, 'updateDetailsCount')

         storageModule.addDetails(name, lastname);

         expect(updateCartItemsCountSpy).to.have.been.called.once
    });
  });

```

I need to spy / stub `updateDetailsCount` when `addDetails` but above code is not working and I cannot figure out why and how to achieve this.  Any help is appreciated! Thanks

**Comment by portyu9** [comment]
Hi @Gmiliuta 

The spy does not work because `addDetails()` calls the module's **local function binding**:

```js
updateDetailsCount(allDetails)
```

but your spy wraps the exported property:

```js
cy.spy(storageModule, 'updateDetailsCount')
```

Those are not the same call site. Wrapping the exported module property does not rewrite the lexical reference already used inside `addDetails()`.

One straightforward refactor is to make the dependency a method on an object and call it through that object:

```js
export const storage = {
  updateDetailsCount(details) {
    // ...
  },

  addDetails({ name, lastname }) {
    const allDetails =
      JSON.parse(localStorage.getItem('personDetails')) || []

    allDetails.push(`${name} ${lastname}`)
    localStorage.setItem('personDetails', JSON.stringify(allDetails))

    storage.updateDetailsCount(allDetails)
  },
}
```

Then the spy observes the same property that `addDetails()` calls:

```js
cy.spy(storage, 'updateDetailsCount')

storage.addDetails({
  name: 'tom',
  lastname: 'jones',
})

expect(storage.updateDetailsCount).to.have.been.calledOnce
```

There are also several unrelated errors in the sample as posted:

- `addDetails()` expects one object, but the test passes two arguments.
- `` `${name} {lastname}` `` is missing `$` before `lastname`.
- `fullname` is undefined.
- the assertion refers to `updateCartItemsCountSpy`, although the declared variable is `updateDetailsCountSpy`.

Those should be corrected independently of the spying issue.

Please mark this answer as accepted if it helped, thank you.
